### PR TITLE
TRIAN-1, TRIAN-2: Enhance Docs and Implement Namespaced Config For Clarity

### PR DESCRIPTION
**Pull request by sifbiri**
## Summary

Some of the changes in this PR aim to improve the documentation to be more comprehensive. Namespaced config support has been introduced for better organization, along with additional keys to increase configurability while maintaining compatibility. The spec checking has been expanded for greater robustness, and the views handling in dev mode has been streamlined when using Vite, as well as adding client-side routing.

## Related Issues
Closes TRIAN-1
Closes TRIAN-2

### CODE REVIEW

File: src/triangulum/config.clj [hunk 1]
```diff
@@ -1,80 +1,95 @@
...
 
+(defn- get-mapped-key-ns
+  "Given a namespaced key, returns the corresponding unnamespaced key."
+  [ns-key]
+  (let [new-ns (-> ns-key
+                   namespace
+                   (clojure.string/split #"\.")
+                   second
+                   keyword)]
+    (or (get ns->un-mapping new-ns) new-ns)))
```
Comment (a-luz) [inline src/triangulum/config.clj:90]: We can remove the `or` function call here, passing `new-ns` as the default value to the function `get`.

e.g: `(get ns->un-mapping new-ns new-ns)`

File: src/triangulum/config.clj [hunk 2]
```diff
@@ -85,16 +100,65 @@
...
+  "Given a namespaced key, returns a vector of unnamespaced keys."
+  [ns-key]
+  [(get-mapped-key-ns ns-key) (-> ns-key (name) (keyword))])
+
+(defn join-un-key
+  "Given a sequence of unnamespaced keys, returns a single namespaced key."
+  ([key-path] (join-un-key key-path {:prefix "triangulum"}))
+  ([key-path {:keys [prefix]}]
+   (let [[n k] key-path
+         n (or (get un->ns-mapping n) n)]
```
Comment (a-luz) [inline src/triangulum/config.clj:122]: Here we can remove the `or` function call, passing the default value to the `get` function.

File: src/triangulum/server.clj [hunk 3]
```diff
@@ -41,47 +67,36 @@
...
-            :requires    [:http-port]}
-   :stop   {:description "Stops the server."}
-   :reload {:description "Reloads a running server."}})
+;;===============================================
+;; Actions
+;;===============================================
 
 #_{:clj-kondo/ignore [:shadowed-var]}
 (defn start-server!
   "FIXME: Write docstring"
```
Comment (a-luz) [inline src/triangulum/server.clj:76]: I think this is a good PR to add a docstring to this function

File: src/triangulum/views.clj [hunk 4]
```diff
@@ -216,7 +247,30 @@
...
+                     window.$RefreshSig$ = () => (type) => type;"]
+               [:script {:type "module"
+                         :src "http://localhost:5173/index.html?html-proxy&index=0.js"}]
+               #_[:script {:type "module" :src "http://localhost:5173/@vite/client"}]
+               #_[:meta {:charset "UTF-8"}] [:link {:rel "icon" :type "image/svg+xml" :href "/vite.svg"}]
+               #_[:meta {:name "viewport" :content "width=device-width, initial-scale=1.0"}]
+               #_[:title "Vite + React + TS + Emotion"]]
+              [:body
+               [:div#app]
+               (client-init nil (:params request) (:session request))])}))
```
Comment (a-luz) [inline src/triangulum/views.clj:273]: Do we still need this commented-out function?

**Review by a-luz — COMMENTED** [review]
It works fine from what I have tested, though my knowledge about triangulum is still limited.

**Comment by lambdatronic** [comment]
# README.org
"If you need to provide a symbol that is bound to a handler function
to figwheel-main, you can use `triangulum.handler/development-app` to
load in your project's handler function from `config.edn` with the
`wrap-reload` middleware added."

- Document that `:cljs-init` and `:js-init` now receive `params` and
  `session`, and `session` now contains `:versionDeployed` and
  `:client-keys`.

# config.example2.edn
- 13: `:triangulum.worker/start-fn` -> `:triangulum.worker/start`
- 14: `:triangulum.worker/stop-fn`  -> `:triangulum.worker/stop`

# build_db.clj
- 11: Remove trailing whitespace.
- 13-16: Align specs into columns.

# config.clj
- 15-22: Align specs into columns.
- 44: "Converts namespaces into their equivalent unnamespaced keys."
- 47: "Convers unnamespaced keys into their equivalent namespaces."
- 49: Remove extra blank line.
- 60: `(some? (namespace k))`
- 65: Align let values into columns in bindings vector.
- 88: `(str/split #"\.")`
- 110: Missing docstring.
- 111: `(s/valid? ::nested-config config)`
- 134: Remove trailing whitespace.
- 158: `(not (namespaced-key? k))`

# config_namespaced_spec.clj
- 14-15: Fix `:req` and `:opt` indentation.

# config_nested_spec.clj
- *: Put a single blank line between each spec.
- 39: `:req-un app-req`

# database.clj
- 6: Put libspecs in alphabetical order.
- 12: Add a blank line after comment.
- 12-16: Align specs into columns.

# email.clj
- 2-5: Put libspecs in alphabetical order.

# git.clj
- 10: Remove trailing whitespace.

# handler.clj
- 26: `[triangulum.config :as config :refer [get-config]]`
- 32: Remove trailing whitespace.
- 34-35: Align specs into columns.
- 34: ::string -> ::config/string
- 35: ::string -> ::config/string
- 167-173: Use this function instead of `create-development-app`:
  ```clojure
  (def development-app
    (create-handler-stack
     (fn [request]
       (let [user-handler (-> (get-config :server :handler)
                              resolve-foreign-symbol)]
         (user-handler request)))
     false
     true))
  ```

# https.clj
- 8: Put libspecs in alphabetical order.
- 14-18: Align specs into columns.

# response.clj
- 2: Fix `:as` indentation.
- 7: Remove trailing whitespace.
- 9: `(s/def ::response-type #{:json :end :transit})`

# server.clj
- 14: Put libspecs in alphabetical order.
- 24: Remove `;; server` line.
- 25-41: Align specs into columns.
- 31: `::string` -> `::config/string`
- 34-38: Fix workers specs:
  ```clojure
  (s/def ::nested-worker     (s/keys :req-un [:start :stop]))

  (s/def ::namespaced-worker (s/keys :req [:triangulum.worker/name
                                           :triangulum.worker/start
                                           :triangulum.worker/stop]))

  (s/def ::workers
    (s/or :map    (s/map-of keyword? ::nested-worker)
          :vector (s/coll-of ::namespaced-worker :kind vector?)))
  ```

# views.clj
- 16: sepc -> spec
- 18: Remove `;; views` line.
- 19-37: Align specs into columns.
- 20: Remove `map?` spec from `s/and`.
- 20-22: Use a simpler spec here:
  ```clojure
  (s/def ::localized-text (s/map-of ::lang string?)
  ```
- 28: Use a more robust spec here:
  ```clojure
  (s/def ::hiccup-element (s/cat :tag   ::hiccup-tag
                                 :attrs (s/? ::hiccup-attrs)
                                 :body  (s/? any?)))
  ```
- 30: `clojure.string/starts-with?` -> `str/starts-with?`
- 38: Remove extra blank line.

**Comment by lambdatronic** [comment]
# README.org
- Replace backticks with tildes globally.
- Remove references to `create-development-app`.
- 159: ~js-init~ -> ~:js-init~
- 159: "builds.In" -> "builds. In"
- 159: There are two client-side entrypoint functions, depending on the language chosen:
  - JS front-ends:   ~<:js-init>.pageInit(params, session)~
  - CLJS front-ends: ~(<:cljs-init> params session)~
- Document that ~:client-keys~ are added to the ~session~ map that is
  passed to the client-side entrypoint function.

# config.example2.edn
- 11: Align "scheduler" above other values in this map.

# config_nested_spec.clj
- 48,69: Put a single blank line between each spec.
- *: Remove extra spaces in `s/def` forms.

# git.clj
- 9: Remove extra blank line.

# response.clj
- 9: `:end` -> `:edn`

# server.clj
- 14: Put libspecs in alphabetical order.
- 24-42: Fix alignment.

**Comment by lambdatronic** [comment]
# README.org
- 98-111: Remove these lines.
- 159-160,166: ~params~ comes before ~session~.
- 165: "pass" -> "assign"

# server.clj
- 40-42: Fix alignment.

**Pull request closed without merging**